core/src/ldml/ldml_processor.cpp

```diff
@@ -252,13 +252,15 @@ ldml_processor::process_backspace(km_core_state *state) const {
 void
 ldml_processor::process_key(km_core_state *state, km_core_virtual_key vk, uint16_t modifier_state) const {
   // Look up the key
-  const std::u16string key_str = keys.lookup(vk, modifier_state);
+  bool found = false;
+  const std::u16string key_str = keys.lookup(vk, modifier_state, found);
 
-  if (key_str.empty()) {
+  if (!found) {
```

Comment:
We need to exclude modifiers (incl. Caps) from this test. Press and release Ctrl or Shift should not invalidate context. Press and release Alt on Windows is a special case -- because it opens the menu. But it may not need to invalidate context because other actions will do that anyway.

However, Ctrl+A _should_ invalidate context. So we need to be aware of the modifier state of key combinations and invalidate context for missing key events. And if we define Ctrl+Q in the keyboard, does that mean that all other Ctrl+key combinations will be treated as gap? Philosophy time.

Also, I wonder how many keyboard developers will get annoyed when their spacebar doesn't work until they add a key for it?

Reply:
- we're only on key down here, for one thing. key up events are silently ignored.  (line 185) - should that change?
- if Ctrl-A was assigned as a key, then it should not invalidate context. So I'm going to need some more details here...
- space - we could detect space here and simply send it to the OS, invalidating context.  But, since we already have an implied `space` key in LDML, kmc could simply slip that one in unless it's otherwise defined. (i.e. instead of gap).  The net effect would be that if you define _no_ keys you'll get a keyboard with one functional key- the spacebar. 
- seems like there's enough here that it's time for a CLDR ticket: https://unicode-org.atlassian.net/browse/CLDR-17262

Reply:
> * we're only on key down here, for one thing. key up events are silently ignored.  (line 185) - should that change?

The key up events don't tend to harm if the key is already up (in my experience). So should be okay to go through.

> * if Ctrl-A was assigned as a key, then it should not invalidate context. So I'm going to need some more details here...

If Ctrl+A is assigned, then yes, it gets processed as normal. If it isn't assigned, however, and falls back through to the app (as a user would expect), doing a select-all on Windows for example, then we would want a context reset.

Reply:
> key up

I don't think you mean "all keyup events trigger passthrough/context reset" as that would be disruptive to markers.   so I'm not sure what the code should do different here.

> ctrl=A 

OK so far.

I'm still unclear on *which* vkeys need to be passed through to the OS, causing a context reset.   is there a list? should the list be in CLDR?